### DISCUSSION TITLE
Run scripts in Controller

### DISCUSSION THREAD

**Discussion by santakdalai90** [argoproj/argo-workflows#12361] [Ideas]
Hi,
We have a way to run scripts in Argo Workflow but all of those require an image in the template and will run a container to execute the script.
https://argoproj.github.io/argo-workflows/workflow-concepts/#script
https://argoproj.github.io/argo-workflows/walk-through/scripts-and-results/

Is there a way to run scripts natively on the workflow-controller container itself, so that for trivial jobs like print etc we don't have to start a fresh container?

If not, then would it be advisable to extend the functionality of `ScriptTemplate`
https://github.paypal.com/PaaS-R/kubeioargocd/blob/bc1872c1a700e08456361ede808464c250d58a4e/argo-workflows/pkg/apis/workflow/v1alpha1/workflow_types.go#L2925
so that there can be scripts which can run native on workflow-controller container itself?

Or, do you see this whole idea as an anti-pattern?

Please advise.
Thanks

**Comment by Joibel** [comment]
I see this as an anti-pattern.

The controller has way to many privileges in the cluster for this to be safe, and there is currently only ever one of them so it would be overloaded. It also really shouldn't actually have a script processor (/bin/bash) on it for security.

Some executor plugin which used hot pooled pods to do the work would be nice for this and other quick work would be nice.

**Reply by santakdalai90** [reply]
@Joibel Thanks for the advice. I would like to know more about hot pooled pods. Would you have any reference for that please?

Thanks

**Reply by Joibel** [reply]
I don't, I just made up an idea when I was responding to you. It would be a nice extension to workflows, but I've not seen that anyone has tried implementing it.

**Reply by Joibel** [reply]
The current executor plugin system could not do this - it would take some controller work to make it happen.

**Reply by agilgur5** [reply]
There is also an existing feature request around this for "Job templates": https://github.com/argoproj/argo-workflows/issues/8786

> The current executor plugin system could not do this

The endpoint plugin type from https://github.com/argoproj/argo-workflows/issues/8142 would make this partially possible as well (it would still spin up an executor)
